File: ds-java/src/heap/Heap.java
package heap;

import java.util.ArrayList;
import java.util.List;

public class Heap {
    private final List<Integer> heap = new ArrayList<>();

    public void insert(int value) {
        heap.add(value);
        if (heap.size() == 1)
            return;
        upHeap(heap.size() - 1);
    }

    public int poll() {
        if (heap.isEmpty()) {
            throw new IllegalStateException("힙이 비어있음");
        }
        var value = heap.get(0);
        if (heap.size() == 1) {
            heap.remove(0);
        } else {
            downHeap(heap.size() - 1);
        }
        return value;
    }

    public int peek() {
        if (heap.isEmpty()) {
            throw new IllegalStateException("힙이 비어있음");
        }
        return heap.get(0);
    }

    private void upHeap(int index) {
        int child = index;
        while (heap.get(child) < heap.get((child - 1) / 2)) {
            int parent = (child - 1) / 2;
            swap(child, parent);
        }
    }

    private void downHeap(int index) {
        heap.set(0, heap.get(index));
        heap.remove(index);

        int parent = 0;
        int size = heap.size();

        while (true) {
            int left = leftChild(parent);
            int right = rightChild(parent);
            int smallest = parent;

            if (left < size && heap.get(left) < heap.get(smallest)) {
                smallest = left;
            }

            if (right < size && heap.get(right) < heap.get(smallest)) {
                smallest = right;
            }

            if (smallest == parent)
                break;

            swap(parent, smallest);
            parent = smallest;
        }
    }

    private int leftChild(int index) {
        return index * 2 + 1;
    }

    private int rightChild(int index) {
        return index * 2 + 2;
    }

    private void swap(int i, int j) {
        int tmp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, tmp);
    }

    @Override
    public String toString() {
        var sb = new StringBuilder();
        int level = 0;
        int count = 0;
        int maxCount = 1;

        for (int i = 0; i < heap.size(); i++) {
            sb.append(heap.get(i)).append(" ");
            count++;

            if (count == maxCount) {
                sb.append("\n");
                level++;
                count = 0;
                maxCount = (int) Math.pow(2, level);
            }
        }

        return sb.toString();
    }
}

File: data-structures/tree/heap/heap.py
class Heap: 
    def __init__(self):
        self.hq = []

    def add(self, num): 
        self.hq.append(num)
        self.__up()

    def pop(self): 
        num = self.hq[0]
        self.hq[0] = self.hq[-1]
        self.hq.pop()
        self.__down()
        return num

    def __up(self) -> None:
        curr = len(self.hq) - 1

        while self.__parent(curr) >= 0 and self.hq[self.__parent(curr)] > self.hq[curr]: 
            self.hq[self.__parent(curr)], self.hq[curr] = self.hq[curr], self.hq[self.__parent(curr)]
            curr = self.__parent(curr)

    def __down(self) -> None: 
        curr = 0 

        last = len(self.hq) - 1

        leftchild = self.__left_child(curr)
        rightchild = self.__right_child(curr)
        
        while leftchild <= last: 
            if rightchild <= last: 
                if self.hq[leftchild] >= self.hq[curr] and self.hq[rightchild] >= self.hq[curr]: 
                    break

                if self.hq[leftchild] < self.hq[rightchild]: 
                    self.hq[leftchild], self.hq[curr] = self.hq[curr], self.hq[leftchild]
                    curr = leftchild
                    leftchild = self.__left_child(curr)
                    rightchild = self.__right_child(curr)
                else: 
                    self.hq[rightchild], self.hq[curr] = self.hq[curr], self.hq[rightchild]
                    curr = rightchild
                    leftchild = self.__left_child(curr)
                    rightchild = self.__right_child(curr)
                continue
            else: 
                if self.hq[leftchild] >= self.hq[curr]: 
                    break
                else:
                    self.hq[leftchild], self.hq[curr] = self.hq[curr], self.hq[leftchild]
                    curr = leftchild
                    leftchild = self.__left_child(curr)
                    rightchild = self.__right_child(curr)


        
    def __parent(self, child) -> int: 
        return (child - 1) // 2
    
    def __left_child(self, parent) -> int: 
        return parent * 2 + 1
    
    def __right_child(self, parent) -> int: 
        return parent * 2 + 2
